# CAMISIM

CAMISIM is a software to model abundance distributions of microbial communities and to simulate corresponding shotgun metagenome datasets.
It was mainly developed for the [Critical Assessment of Metagenome Annotation (CAMI)](http://microbiome-cosi.org/cami) challenge, but should be suitable for general use. Please don't hesitate to [open a new issue](https://github.com/CAMI-challenge/CAMISIM/issues) if you run into problems or need help.

### Documentation 
* [User manual](https://github.com/CAMI-challenge/CAMISIM/wiki/User-manual)
* [Configuration File Options](https://github.com/CAMI-challenge/CAMISIM/wiki/Configuration-File-Options)
* [File Formats](https://github.com/CAMI-challenge/CAMISIM/wiki/File-Formats)

### Citation

If you use CAMISIM, please cite the publication at *Microbiome*:
* Fritz*, Hofmann*, et al. (2019). **CAMISIM: Simulating metagenomes and microbial communities.** *Microbiome*, 2019, 7:17. doi:[10.1186/s40168-019-0633-6](https://doi.org/10.1186/s40168-019-0633-6)

A part of CAMISIM's functionality was also described in the CAMI manuscript, thus you may also cite:
* Sczyrba*, Hofmann*, Belmann*, et al. (2017). **Critical Assessment of Metagenome Interpretation—a benchmark of metagenomics software.** *Nature Methods*, 14, 11:1063–1071. doi:[10.1038/nmeth.4458](https://doi.org/10.1038/nmeth.4458)
